File: TP_MPI/pi.c
/* 
   Programme de calcul du pi par intégration du quart de cercle
   Fichier à compiler avec mpicc
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <math.h>
#include <mpi.h>

// Fonction d'intégration du cercle pour l'intervalle de x [deb : fin] avec n trapèzes
double integration(long n, double deb, double fin);

//
// Programme principal
//
int main(int argc, char **argv)
{
  int    opt;           // Variable pour les options
  double cdeb, cfin;    // Chronos de début et fin
  double pi = 0;        // Valeur finale de PI
  long    n = 10000000; // Nombre total de trapèzes

  int    nbP, num;      // Nombre de processus et numéro du processus courant
  long   nLoc;          // Nombre de trapèzes sur le processus courant
  double pi_part;       // Valeur partielle de PI
  int    reste;         // Reste du nombre de trapèzes sur le nombre de processus
  double deb, fin;      // Intervalle de x traité par le processus courant

  // Gestion des options du programme en ligne de commande  
  while ((opt = getopt(argc, argv, "n:")) != -1) {
    switch (opt) {
    case 'n':
      n = atol(optarg);
      break;
    }
  }

  // Initialisation du système MPI
  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nbP); // Récupération du nombre de processus
  MPI_Comm_rank(MPI_COMM_WORLD, &num); // Récupération du numéro du processus courant

  // Démarrage chrono début
  cdeb = MPI_Wtime();

  // Calcul des indices et nombre de trapèzes pour le processus courant
  nLoc  =   n; // À MODIFIER ...
  reste =   0; // À MODIFIER ...
  fin   = 1.0; // À MODIFIER ...
  deb   = 0.0; // À MODIFIER ...

  // Intégration du quart de cercle dans l'intervalle de x [0.0, 1.0] avec n trapèzes
  pi_part = integration(nLoc, deb, fin); 

  // Affichage des infos de calcul de chaque processus
  printf("Le proc %d traite %ld trapèzes de %f à %f et obtient %.15lf\n", num, nLoc, deb, fin, pi_part);

  // Fusion du résultat sur le processus 0
  // ... À MODIFIER ...
  pi = pi_part;

  // Récupération chrono fin
  cfin = MPI_Wtime();

  // Affichage du résultat par le processus 0
  if(num == 0){
    printf("Valeur de pi                     : %1.15f\n", pi);
    printf("Écart à la valeur standard de pi : %.3e\n", M_PI - pi);
    printf("Calculée en                      : %fs\n", cfin - cdeb);
  }

  // Arrêt du système MPI
  MPI_Finalize();

  return(EXIT_SUCCESS);
}

//
// Fonction du cercle de rayon 1 sur l'intervalle [-1,1]
//
double cercle(double x)
{
  return sqrt(1 - x * x);
}

//
// Intégration du cercle pour l'intervalle de x [deb : fin] avec n trapèzes
//
double integration(long n, double deb, double fin)
{
  long i;                                           // Indice de boucle
  double x;                                         // valeur de x dans l'intervalle
  double h = (fin - deb) / n;                       // Largeur des trapèzes
  double somme = (cercle(deb) + cercle(fin)) / 2.0; // Contributions des deux points extrêmes

  for (i = 1; i < n; i++){ // Contributions des points internes de l'intervalle
    x = deb + h * i;
    somme += cercle(x);
  }

  return 4.0 * h * somme;
}

File: TP_MPI/pi.py
#!/bin/python3
# -*- coding: utf-8 -*-

# Programme qui calcule une approximation de PI avec MPI
# Execution : invocation directe dans un terminal ou via python3

# Bibliothèques utilisées
import sys
import math

# Import de MPI
from mpi4py import MPI

#
# Équation du cercle unitaire
#
def cercle(x):
    return math.sqrt(1.0 - x * x)

#
# Intégration partielle du quart de disque positif à partir de la position debut avec nLoc trapèzes de largeur dx
#
def integration(nLoc, debut, dx):
    res = (cercle(debut) + cercle(debut + nLoc * dx)) / 2.0 # Contributions des points aux extrémités

    x = debut                # Position x de départ de l'intégration partielle
    for i in range(1, nLoc): # Contributions des points internes
        x   += dx
        res += cercle(x);

    return 4.0 * dx * res;

#
# Fonction principale
#
def main():

    # Valeur par défaut du nombre de trapèzes
    n       = 1000000 # Nombre de trapèzes
    
    # Paramètres de la ligne de commande
    argc = len(sys.argv)
    for i in range(argc):
        if sys.argv[i] == "-n":
            i = i + 1
            n = int(sys.argv[i])
    
    # Variables internes du programme
    pi      = 0.0     # Valeur finale de PI
    pi_part = 0.0     # Valeur partielle de PI calculée par le processus courant
    tdeb    = 0.0     # Temps de début du calcul
    tfin    = 0.0     # Temps de fin du calcul
    debut   = 0.0     # Position x du 1er trapèze de l'intégration partielle
    dx      = 1.0 / n # Largeur des trapèzes
    nLoc    = n       # Nombre de trapèzes locaux par processus (mis à jour ci-dessous)

    # Initialisation MPI
    comm = MPI.COMM_WORLD  # Groupe de communication
    idP  = comm.Get_rank() # Identifiant du processus
    nbP  = comm.Get_size() # Nombre de processus dans le groupe

    # Temps de début du calcul
    tdeb = MPI.Wtime()

    # Calcul des paramètres locaux : nb trapèzes et indices
    nLoc   = n   # !! À MODIFIER !!
    reste  = 0   # !! À MODIFIER !!
    debut  = 0.0 # !! À MODIFIER !!

    # Integration partielle du quart de disque supérieur-droit à partir de la position debut avec nLoc trapèzes de largeur dx
    pi_part = integration(nLoc, debut, dx)

    # Affichage des résultats partiels
    print("Le processus %d intègre %d trapèzes de largeur %f depuis la position %f : %.15f" % (idP, nLoc, dx, debut, pi_part))

    # Fusion globale des résultats partiels
    pi = pi_part # !! À MODIFIER !!

    # Temps de fin du calcul
    tfin = MPI.Wtime()
    
    # Affichage du résulat global par le processus 0
    if (idP == 0):
        print("Approximation de PI             : %1.15f" % (pi))
        print("Différence avec valeur standard : %.3e" % (math.pi - pi))
        print("Calcul en                       : %fs" % (tfin - tdeb))

#
# Programme principal
#
if  __name__  == "__main__":
    main()
